File: algo/215. Kth Largest Element in an Array.js
// @9 Array

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findKthLargest = function(nums, k) {
    var pivot;
    // might should shuffle the array?
    if (nums.length > 2) {
        pivot = nums[2];
    } else {
        pivot = nums[0];
    }
    var left = [];
    var mid = [];
    var right = [];
    for (var num of nums) {
        if (num < pivot) {
            left.push(num);
        } else if (num == pivot) {
            mid.push(num);
        } else {
            right.push(num);
        }
    }
    if (k <= right.length) {
        return findKthLargest(right, k);
    } else if (k - right.length <= mid.length) {
        return pivot;
    } else {
        return findKthLargest(left, k-mid.length-right.length);
    }
};

File: round2/215. Kth Largest Element in an Array.cs
public class Solution {
    public int FindKthLargest(int[] nums, int k) {
        int pivot;
        if (nums.Length > 2) pivot = nums[2];
        else pivot = nums[0];
        var left = new List<int>();
        var mid = new List<int>();
        var right = new List<int>();
        foreach (var num in nums) {
            if (num < pivot) left.Add(num);
            else if (num > pivot) right.Add(num);
            else mid.Add(num);
        }
        if (k <= right.Count) {
            return FindKthLargest(right.ToArray(), k);
        } else if (k - right.Count <= mid.Count) {
            return pivot;
        } else {
            return FindKthLargest(left.ToArray(), k-mid.Count-right.Count);
        }
    }
}